### PR TITLE
Improve performance of tuple operations

### PR DESCRIPTION
**Pull request by brunnerant**
In this pull request, I made some extensive benchmarks of the current tuple operations (cons, concat, tail, map, zip, ...). I analyzed their performance and tried to understand what were the bottlenecks of the current implementation, and improved them.

Not surprisingly, it turns out that the most efficient way to implement them is to avoid conversions between tuple, iterators and arrays as much as possible because this is the main source of overhead.

For Tuple1 to Tuple22, making a big match on the tuple arity is what gives the best performance (but very ugly code unfortunately).
For TupleXXL, performing the operation on the array seems to be the way to go.

See @nicolasstucki, my semester project supervisor, for more information.

### CODE REVIEW

**Comment by brunnerant** [comment]
test performance with #tuple-runtime : 0a33ad34e

**Comment by brunnerant** [comment]
`dotty.tools.dotc.CompilationTests.tastyBootstrap` doesn't pass, but all the other tests do ? Has anyone an idea of what could be causing this (because I'm not familiar with tasty at all) ?

**Comment by nicolasstucki** [comment]
You may have found a bug in the compiler. I would suggest using git bisect to find the commit where it started failing.

**Comment by brunnerant** [comment]
Ok, I will do this.

**Comment by brunnerant** [comment]
I identified the commit where it starts failing, it is e1a01d2. It probably has something to do with polymorphic lambdas so I will try to see if I can fix it.

**Comment by brunnerant** [comment]
Ok, the problem was coming from `dynamicMap`. I was pattern matching using existential types, and then applying the polymorphic lambda to each tuple element. Maybe the compiler cannot infer the types correctly in this case, so what I did for now is that I cast everything to `Object`.

**Comment by brunnerant** [comment]
Should we try to test the performance using the profile `tuple-runtime`, comparing against master ?

**Comment by nicolasstucki** [comment]
test performance with #tuple-runtime

**Comment by nicolasstucki** [comment]
@liufengyun did the benchmark run?

**Comment by liufengyun** [comment]
There is a bug report https://github.com/lampepfl/bench/issues/47 . However, I couldn't find any logs. I just rescheduled the job.

**Comment by liufengyun** [comment]
Currently, it does not support profile names with a hyphen.

```
unknown profile: tuple
```

I renamed `tuple-runtime` to `tupleRuntime`, now the job is running.

**Comment by brunnerant** [comment]
Ok, thanks for fixing it, I didn't know about the hyphen limitation !
Is there a way to launch the benchmarks on the master branch ?

File: library/src/scala/runtime/DynamicTuple.scala [hunk 1]
```diff
@@ -24,6 +24,28 @@ object DynamicTuple {
     elems1
   }
 
+  def dynamicToArray(self: Tuple): Array[Object] = (self: Any) match {
+    case self: Unit => Array.emptyObjectArray
+    case self: TupleXXL => self.toArray
+    case self: Product => productToArray(self)
+  }
```
Comment (nicolasstucki) [inline library/src/scala/runtime/DynamicTuple.scala]: The least common case is `Unit` which can be put at the end.
```scala
  def dynamicToArray(self: Tuple): Array[Object] = (self: Any) match {
    case self: TupleXXL => self.toArray
    case self: Product => productToArray(self)
    case self: Unit => Array.emptyObjectArray
  }
```

File: library/src/scala/runtime/DynamicTuple.scala [hunk 2]
```diff
@@ -24,6 +24,28 @@ object DynamicTuple {
...
+
+  def dynamicToIArray(self: Tuple): IArray[Object] = (self: Any) match {
+    case self: Unit => Array.emptyObjectArray.asInstanceOf[IArray[Object]] // TODO use IArray.emptyObjectIArray
+    case self: TupleXXL => self.elems
+    case self: Product => productToArray(self).asInstanceOf[IArray[Object]]
```
Comment (nicolasstucki) [inline library/src/scala/runtime/DynamicTuple.scala]: Put the `Unit` case at the end

**Comment by nicolasstucki** [comment]
@liufengyun could you execute these benchmarks on master?

**Comment by brunnerant** [comment]
> @liufengyun could you execute these benchmarks on master?

@liufengyun  Is it possible to run them on both branches, to have a comparison ? Thanks !

**Comment by liufengyun** [comment]
Now they are added to the default profile: https://github.com/lampepfl/bench/commit/2acb4ef7ced063f4a9ec5a0269ca9a1fb528d503

**Comment by nicolasstucki** [comment]
test performance with #tupleRuntime: master 08613cba238d1bf97fd8826be51ceae88b40252f

**Pull request merged**